packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts

```diff
@@ -0,0 +1,52 @@
+/**
+ * External dependencies
+ */
+import { useDebounce } from '@wordpress/compose';
+import { useCallback } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { SelectControlProps } from '../select-control';
+
+export const DEFAULT_DEBOUNCE_TIME = 250;
+
+export default function useAsyncFilter< T >( {
+	filter,
+	onFilterStart,
+	onFilterEnd,
+	onFilterError,
+	onInputChange,
+	debounceTime = DEFAULT_DEBOUNCE_TIME,
+	...props
+}: UseAsyncFilterInput< T > ): SelectControlProps< T > {
+	const handleInputChange = useCallback(
+		function handleInputChangeCallback( value?: string ) {
+			if ( onFilterStart ) onFilterStart( value );
+			filter( value )
+				.then( ( filteredItems ) => {
+					if ( onFilterEnd ) onFilterEnd( filteredItems, value );
+				} )
+				.catch( ( error: Error ) => {
+					if ( onFilterError ) onFilterError( error, value );
+				} )
+				.finally( () => {
+					if ( onInputChange ) onInputChange( value, {} );
+				} );
+		},
+		[ filter, onFilterStart, onFilterEnd, onFilterError, onInputChange ]
+	);
+
+	return {
+		...props,
```

Comment:
Somewhat related, I was thinking about whether we should passthrough the props here or have the consumer separately set them...

```
<SelectControl< DefaultItemType > { ...selectProps } { ...asyncSelectProps }>
```

But, if we want to be more opinionated as Josh suggests, it could make sense to keep them being passed through as is currently done to allow for that.

Reply:
> - a suffix (a spinner icon) when we're fetching
> - items for each state (loading vs finished)

I think this can be implemented in a different issue.  Since it requires some design decisions from @jarekmorawski. 

> - a default getFilteredItems which just returns all in items fetched

From my point of view this function should be removed and use the sync version of this hook instead, so it will be a code that is gonna be deleted if I add a default value to getFilteredItems inside the useAsyncFilter hook.

Reply:
That sounds good to me.